### DISCUSSION TITLE
[feature request] Support CONFIG_PREEMPT_RT format

### DISCUSSION THREAD

**Discussion by 7-RED** [google/perfetto#947] [User Ideas]
The format printed under some configurations cannot be parsed on Perfetto. It is hoped that in the future, supported formats can be added along with kernel version updates.
https://github.com/torvalds/linux/blob/v5.15/kernel/trace/trace_osnoise.c

```
static void print_osnoise_headers(struct seq_file *s)
{
	if (osnoise_data.tainted)
		seq_puts(s, "# osnoise is tainted!\n");

	seq_puts(s, "#                                _-------=> irqs-off\n");
	seq_puts(s, "#                               / _------=> need-resched\n");
	seq_puts(s, "#                              | / _-----=> need-resched-lazy\n");
	seq_puts(s, "#                              || / _----=> hardirq/softirq\n");
	seq_puts(s, "#                              ||| / _---=> preempt-depth\n");
	seq_puts(s, "#                              |||| / _--=> preempt-lazy-depth\n");
	seq_puts(s, "#                              ||||| / _-=> migrate-disable\n");

	seq_puts(s, "#                              |||||| /          ");
	seq_puts(s, "                                     MAX\n");

	seq_puts(s, "#                              ||||| /                         ");
	seq_puts(s, "                    SINGLE      Interference counters:\n");

	seq_puts(s, "#                              |||||||               RUNTIME   ");
	seq_puts(s, "   NOISE  %% OF CPU  NOISE    +-----------------------------+\n");

	seq_puts(s, "#           TASK-PID      CPU# |||||||   TIMESTAMP    IN US    ");
	seq_puts(s, "   IN US  AVAILABLE  IN US     HW    NMI    IRQ   SIRQ THREAD\n");

	seq_puts(s, "#              | |         |   |||||||      |           |      ");
	seq_puts(s, "       |    |            |      |      |      |      |      |\n");
}
#else /* CONFIG_PREEMPT_RT */
static void print_osnoise_headers(struct seq_file *s)
{
	if (osnoise_data.tainted)
		seq_puts(s, "# osnoise is tainted!\n");

	seq_puts(s, "#                                _-----=> irqs-off\n");
	seq_puts(s, "#                               / _----=> need-resched\n");
	seq_puts(s, "#                              | / _---=> hardirq/softirq\n");
	seq_puts(s, "#                              || / _--=> preempt-depth     ");
	seq_puts(s, "                       MAX\n");

	seq_puts(s, "#                              || /                         ");
	seq_puts(s, "                    SINGLE      Interference counters:\n");

	seq_puts(s, "#                              ||||               RUNTIME   ");
	seq_puts(s, "   NOISE  %% OF CPU  NOISE    +-----------------------------+\n");

	seq_puts(s, "#           TASK-PID      CPU# ||||   TIMESTAMP    IN US    ");
	seq_puts(s, "   IN US  AVAILABLE  IN US     HW    NMI    IRQ   SIRQ THREAD\n");

	seq_puts(s, "#              | |         |   ||||      |           |      ");
	seq_puts(s, "       |    |            |      |      |      |      |      |\n");
}
#endif /* CONFIG_PREEMPT_RT */
```

**Comment by LalitMaganti** [comment]
Patches for this welcome. Very unlikely we will get to it ourselves.

**Comment by 7-RED** [comment]
> Patches for this welcome. Very unlikely we will get to it ourselves.

Why not? Is it because of the logic of preempt?

**Comment by LalitMaganti** [comment]
It's just not a priority for us. We have a lot commitments already going on and we only prioritise features from the open source community when they are going to affect many of our users.

This is the first time this feature has been requested in the 7 years of Perfetto's existence so is just not a priority.

**Comment by 7-RED** [comment]
Okay I know！But I need to study source code or you can tell me roughly which parts I should look at and make modifications, and I can make a patch.

I think someone will mention this feature in the future.

**Comment by LalitMaganti** [comment]
> Okay I know！But I need to study source code or you can tell me roughly which parts I should look at and make modifications, and I can make a patch.

Well this looks to me like an entirely new file format. Even the if CONFIG_PREEMPT_RT is disabled, we do not parse that.

So you need to make a bunch of changes throughout the codebase to make this happen:
1) Adding a new trace type at [1] and [2] and [3]
2) Adding a new folder in [4] and add an implementation of [5]  to tokenize each line
3) Plumb the data through the sorter [6]
4) Add a parser to the folder you added in [4] to receive the data from the sorter and then add it to the appropriate tables (to me it looks like you will need to add it to the counter table)

Instructions on contributing in general can be found at https://perfetto.dev/docs/contributing/getting-started

You might end up with something similar to similar to https://r.android.com/3368462 where someone from the community added support for a new trace type.

So it's not some simple task, it requires quite a bit of work :)

> I think someone will mention this feature in the future.

Maybe but unless it's a priority for many users OR it's a priority for some team we explicitly support inside Google, we won't get to this.

[1] https://github.com/google/perfetto/blob/master/src/trace_processor/util/trace_type.cc#L109
[2] https://github.com/google/perfetto/blob/master/src/trace_processor/util/trace_type.h
[3] https://github.com/google/perfetto/blob/master/src/trace_processor/trace_reader_registry.cc
[4] https://github.com/google/perfetto/tree/master/src/trace_processor/importers
[5] https://github.com/google/perfetto/blob/master/src/trace_processor/importers/common/chunked_trace_reader.h
[6] https://github.com/google/perfetto/blob/master/src/trace_processor/sorter/trace_sorter.h

**Comment by 7-RED** [comment]
Thank Lalit Maganti, I will evaluate my resources and then start developing this patch.
